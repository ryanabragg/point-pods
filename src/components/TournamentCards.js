import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';

import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';

const styles = theme => ({
  root: {
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
  },
  card: {
    minWidth: 275,
    margin: theme.spacing.unit,
  },
  right: {
    marginLeft: 'auto',
  },
});

class TournamentCards extends Component {
  render() {
    const { classes, displayOnly, tournaments, status, category, search } = this.props;
    let cards;
    switch(status) {
      default:
      case 'Show All': cards = tournaments; break;
      case 'Pending': cards = tournaments.filter(t => t.staging); break;
      case 'In Progress': cards = tournaments.filter(t => !t.staging && !t.done); break;
      case 'Incomplete': cards = tournaments.filter(t => !t.done); break;
      case 'Complete': cards = tournaments.filter(t => t.done); break;
    }
    if(category)
      cards = cards.filter(t => t.category == category);
    if(search)
      cards = cards.filter(t => (
        t.name.toLowerCase().includes(search.toLowerCase())
        || t.category.toLowerCase().includes(search.toLowerCase())
        || t.description.toLowerCase().includes(search.toLowerCase())
      ));
    return (
      <div className={classes.root}>
        {cards.map(({id, name, category, done, staging, players}) => (
          <Card key={id} id={id} className={classes.card}>
            <CardHeader
              title={name}
              subheader={category}
            />
            <CardContent>
              <Typography>
                {`${players.length} Total Players`}
              </Typography>
              <Typography>
                {`${players.filter(t => !t.dropped).length} Active Players`}
              </Typography>
            </CardContent>
            {!displayOnly && (
              <CardActions>
                <Button size='small'
                  className={classes.right}
                  href={`/${staging ? 'new' : 'tournament'}/${id}`}
                >
                  {done ? 'Details' : 'Continue'}
                </Button>
              </CardActions>
            )}
          </Card>
        ))}
        {!cards.length && (
          <Card id='not-found' className={classes.card}>
            <CardHeader
              title='No Tournaments Found'
            />
          </Card>
        )}
      </div>
    );
  }
}

TournamentCards.defaultProps = {
  displayOnly: false,
  tournaments: [],
  selected: [],
  onSelect: (tournaments) => null,
  status: 'Show All',
  category: '',
  search: '',
};

TournamentCards.propTypes = {
  classes: PropTypes.object.isRequired, // added by withStyles
  displayOnly: PropTypes.bool,
  tournaments: PropTypes.array,
  selected: PropTypes.array,
  onSelect: PropTypes.func,
  status: PropTypes.oneOf([
    'Show All',
    'Pending',
    'In Progress',
    'Incomplete',
    'Complete',
  ]),
  category: PropTypes.string,
  search: PropTypes.string,
};

export default withStyles(styles, { withTheme: true })(TournamentCards);
