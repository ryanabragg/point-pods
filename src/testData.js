import api from './api';

const testSettings = {
  pairingMethod: api.pairingMethods.HISTORIC,
  pairingMethodInitial: api.pairingMethods.HISTORIC,
  podSizeMinimum: 4,
  podSizeMaximum: 7,
};

const testPlayers = [{
  id: 'mock0',
  name: 'test0',
  points: 0,
}, {
  id: 'mock1',
  name: 'test1',
  points: 1,
}, {
  id: 'mock2',
  name: 'test2',
  points: 2,
}, {
  id: 'mock3',
  name: 'test3',
  points: 3,
}, {
  id: 'mock4',
  name: 'test4',
  points: 4,
}, {
  id: 'mock5',
  name: 'test5',
  points: 5,
}, {
  id: 'mock6',
  name: 'test6',
  points: 6,
}, {
  id: 'mock7',
  name: 'test7',
  points: 7,
}, {
  id: 'mock8',
  name: 'test8',
  points: 8,
}, {
  id: 'mock9',
  name: 'test9',
  points: 9,
}];

const testTournaments = [{
  id: 'mock0',
  name: 'test0',
  category: 'cat0',
  description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris vitae tellus nunc. Etiam urna tortor, laoreet vitae commodo porttitor, mattis id purus.',
  date: new Date('2018-01-01'),
  done: false,
  staging: true,
  pairingMethod: api.pairingMethods.RANDOM,
  pairingMethodInitial: api.pairingMethods.RANDOM,
  podSizeMinimum: 3,
  podSizeMaximum: 4,
  rounds: 1,
  players: [{
    id: 'mock0',
    name: 'test0',
    points: 3,
    participated: true,
    dropped: true,
    1: { pod: 1, points: 3 },
  }, {
    id: 'mock1',
    name: 'test1',
    points: 1,
    participated: true,
    dropped: false,
    1: { pod: 1, points: 1 },
  }, {
    id: 'mock2',
    name: 'test2',
    points: 2,
    participated: true,
    dropped: false,
    1: { pod: 1, points: 2 },
  }, {
    id: 'mock3',
    name: 'test3',
    points: 0,
    participated: false,
    dropped: false,
  }],
}, {
  id: 'mock1',
  name: 'test1',
  category: 'cat1',
  description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed molestie suscipit suscipit. Nullam sollicitudin tempus pretium.',
  date: new Date('2018-02-01'),
  done: true,
  staging: false,
  pairingMethod: api.pairingMethods.POINTS,
  pairingMethodInitial: api.pairingMethods.RANDOM,
  podSizeMinimum: 3,
  podSizeMaximum: 5,
  rounds: 2,
  players: [{
    id: 'mock0',
    name: 'test0',
    points: 0,
    participated: true,
    dropped: true,
    1: { pod: 1, points: 0 },
    2: { pod: 1, points: 0 },
  }, {
    id: 'mock1',
    name: 'test1',
    points: 1,
    participated: true,
    dropped: true,
    1: { pod: 1, points: 1 },
    2: { pod: 2, points: 1 },
  }, {
    id: 'mock2',
    name: 'test2',
    points: 2,
    participated: true,
    dropped: false,
    1: { pod: 1, points: 2 },
    2: { pod: 1, points: 2 },
  }, {
    id: 'mock3',
    name: 'test3',
    points: 3,
    participated: true,
    dropped: false,
    1: { pod: 2, points: 3 },
    2: { pod: 2, points: 3 },
  }, {
    id: 'mock4',
    name: 'test4',
    points: 4,
    participated: true,
    dropped: false,
    1: { pod: 2, points: 4 },
    2: { pod: 1, points: 4 },
  }, {
    id: 'mock5',
    name: 'test5',
    points: 5,
    participated: true,
    dropped: false,
    1: { pod: 2, points: 5 },
    2: { pod: 2, points: 5 },
  }, {
    id: 'mock6',
    name: 'test6',
    points: 6,
    participated: true,
    dropped: false,
    2: { pod: 2, points: 6 },
  }],
}, {
  id: 'mock2',
  name: 'test2',
  category: 'cat0',
  description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris vitae tellus nunc. Sed ullamcorper metus vitae nisi rhoncus placerat.',
  date: new Date('2018-01-01'),
  done: false,
  staging: false,
  pairingMethod: api.pairingMethods.RANDOM,
  pairingMethodInitial: api.pairingMethods.RANDOM,
  podSizeMinimum: 3,
  podSizeMaximum: 4,
  rounds: 1,
  players: [{
    id: 'mock0',
    name: 'test0',
    points: 0,
    participated: true,
    dropped: false,
    1: { pod: 1, points: 0 },
  }, {
    id: 'mock1',
    name: 'test1',
    points: 1,
    participated: true,
    dropped: false,
    1: { pod: 1, points: 1 },
  }, {
    id: 'mock2',
    name: 'test2',
    points: 2,
    participated: true,
    dropped: false,
    1: { pod: 1, points: 2 },
  }, {
    id: 'mock3',
    name: 'test3',
    points: 3,
    participated: true,
    dropped: false,
    1: { pod: 1, points: 3 },
  }],
}];

const mockRoute = {
  history: {
    push: jest.fn(to => to),
    replace: jest.fn(to => to),
  },
  match: {
    params: {},
  },
  matchWithInvalid: {
    params: {
      id: 'invalid',
    },
  },
  matchWithID0: {
    params: {
      id: 'mock0',
    },
  },
  matchWithID1: {
    params: {
      id: 'mock1',
    },
  },
};

let mockStore = {};
const mockStorage = {
  setItem: jest.fn((key, val) => Promise.resolve((mockStore[key] = val))),
  getItem: jest.fn(key => Promise.resolve(mockStore[key])),
  removeItem: jest.fn(key => { delete mockStore[key]; Promise.resolve(); }),
  clear: jest.fn(() => Promise.resolve((mockStore = {}))),
  keys: jest.fn(() => Promise.resolve(Object.keys(mockStore))),
  length: jest.fn(() => Promise.resolve(Object.keys(mockStore).length)),
};

let defaultStore = null;
function mockAPI(api, storage) {
  defaultStore = api.store;
  api.store = storage;
  mockStore = {
    settings: testSettings,
    players: testPlayers,
    tournaments: testTournaments,
  };
}
function demockAPI(api) {
  api.store = defaultStore;
  defaultStore = null;
}

export {
  testSettings,
  testPlayers,
  testTournaments,
  mockRoute,
  mockStore,
  mockStorage,
  mockAPI,
  demockAPI,
};
