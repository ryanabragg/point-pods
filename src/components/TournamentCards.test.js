import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Card from '@material-ui/core/Card';

import TournamentCards from './TournamentCards';

import { testTournaments } from '../testData';

test('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<TournamentCards />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test('matches the prior snapshot', () => {
  const component = renderer.create(<TournamentCards />);
  expect(component.toJSON()).toMatchSnapshot();
  component.update(<TournamentCards tournaments={testTournaments} />);
  expect(component.toJSON()).toMatchSnapshot();
});

test('selection');

describe('filter props', () => {
  test('status', () => {
    const wrapper = shallow(<TournamentCards tournaments={testTournaments} />);
    const component = wrapper.dive();
    expect(component.find(Card).length).toBe(testTournaments.length);
    component.setProps({
      category: '',
      status: 'Pending',
    });
    expect(component.find(Card).length).toBe(1);
    expect(component.find(Card).prop('id')).toBe('mock0');
    component.setProps({
      status: 'In Progress',
    });
    expect(component.find(Card).length).toBe(1);
    expect(component.find(Card).prop('id')).toBe('mock2');
    component.setProps({
      status: 'Incomplete',
    });
    expect(component.find(Card).length).toBe(2);
    expect(component.find(Card).at(0).prop('id')).toBe('mock0');
    expect(component.find(Card).at(1).prop('id')).toBe('mock2');
    component.setProps({
      status: 'Complete',
    });
    expect(component.find(Card).length).toBe(1);
    expect(component.find(Card).prop('id')).toBe('mock1');
  });

  test('category', () => {
    const wrapper = shallow(<TournamentCards tournaments={testTournaments} />);
    const component = wrapper.dive();
    expect(component.find(Card).length).toBe(testTournaments.length);
    component.setProps({
      category: 'cat0',
    });
    expect(component.find(Card).length).toBe(2);
    expect(component.find(Card).at(0).prop('id')).toBe('mock0');
    expect(component.find(Card).at(1).prop('id')).toBe('mock2');
    component.setProps({
      category: 'cat1',
    });
    expect(component.find(Card).length).toBe(1);
    expect(component.find(Card).prop('id')).toBe('mock1');
  });

  test('search', () => {
    const search = [{
      term: 'test2',
      expect: [ 'mock2' ],
    }, {
      term: 'dolor sit amet',
      expect: [ 'mock0', 'mock1', 'mock2' ],
    }, {
      term: 'mauris vitae',
      expect: [ 'mock0', 'mock2' ],
    }, {
      term: 'suscipit',
      expect: [ 'mock1' ],
    }];
    const wrapper = shallow(<TournamentCards tournaments={testTournaments} />);
    const component = wrapper.dive();
    expect(component.find(Card).length).toBe(testTournaments.length);
    search.forEach((find, index) => {
      component.setProps({
        search: find.term,
      });
      expect(component.find(Card).length).toBe(find.expect.length);
      component.find(Card).forEach((node, i) => {
        expect(node.prop('id')).toBe(find.expect[i]);
      });
    });
  });

  test('no results', () => {
    const wrapper = shallow(<TournamentCards tournaments={testTournaments} />);
    const component = wrapper.dive();
    expect(component.find(Card).length).toBe(testTournaments.length);
    component.setProps({
      category: 'not-found'
    });
    expect(component.find(Card).length).toBe(1);
    expect(component.find(Card).prop('id')).toBe('not-found');
  });
});
