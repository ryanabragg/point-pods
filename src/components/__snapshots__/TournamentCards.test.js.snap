// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the prior snapshot 1`] = `
<div
  className="TournamentCards-root-63"
>
  <div
    className="MuiPaper-root-67 MuiPaper-elevation1-70 MuiPaper-rounded-68 MuiCard-root-66 TournamentCards-card-64"
    id="not-found"
  >
    <div
      className="MuiCardHeader-root-94"
    >
      <div
        className="MuiCardHeader-content-97"
      >
        <span
          className="MuiTypography-root-100 MuiTypography-headline-105 MuiCardHeader-title-98"
        >
          No Tournaments Found
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`matches the prior snapshot 2`] = `
<div
  className="TournamentCards-root-63"
>
  <div
    className="MuiPaper-root-67 MuiPaper-elevation1-70 MuiPaper-rounded-68 MuiCard-root-66 TournamentCards-card-64"
    id="mock0"
  >
    <div
      className="MuiCardHeader-root-94"
    >
      <div
        className="MuiCardHeader-content-97"
      >
        <span
          className="MuiTypography-root-100 MuiTypography-headline-105 MuiCardHeader-title-98"
        >
          test0
        </span>
        <span
          className="MuiTypography-root-100 MuiTypography-body1-109 MuiTypography-colorTextSecondary-123 MuiCardHeader-subheader-99"
        >
          cat0
        </span>
      </div>
    </div>
    <div
      className="MuiCardContent-root-125"
    >
      <p
        className="MuiTypography-root-100 MuiTypography-body1-109"
      >
        4 Total Players
      </p>
      <p
        className="MuiTypography-root-100 MuiTypography-body1-109"
      >
        3 Active Players
      </p>
    </div>
    <div
      className="MuiCardActions-root-126"
    >
      <a
        className="MuiButtonBase-root-152 MuiButton-root-128 MuiButton-text-130 MuiButton-flat-133 MuiButton-sizeSmall-149 TournamentCards-right-65 MuiCardActions-action-127"
        href="/tournament/mock0"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <span
          className="MuiButton-label-129"
        >
          Continue
        </span>
        <span
          className="MuiTouchRipple-root-155"
        />
      </a>
    </div>
  </div>
  <div
    className="MuiPaper-root-67 MuiPaper-elevation1-70 MuiPaper-rounded-68 MuiCard-root-66 TournamentCards-card-64"
    id="mock1"
  >
    <div
      className="MuiCardHeader-root-94"
    >
      <div
        className="MuiCardHeader-content-97"
      >
        <span
          className="MuiTypography-root-100 MuiTypography-headline-105 MuiCardHeader-title-98"
        >
          test1
        </span>
        <span
          className="MuiTypography-root-100 MuiTypography-body1-109 MuiTypography-colorTextSecondary-123 MuiCardHeader-subheader-99"
        >
          cat1
        </span>
      </div>
    </div>
    <div
      className="MuiCardContent-root-125"
    >
      <p
        className="MuiTypography-root-100 MuiTypography-body1-109"
      >
        8 Total Players
      </p>
      <p
        className="MuiTypography-root-100 MuiTypography-body1-109"
      >
        6 Active Players
      </p>
    </div>
    <div
      className="MuiCardActions-root-126"
    >
      <a
        className="MuiButtonBase-root-152 MuiButton-root-128 MuiButton-text-130 MuiButton-flat-133 MuiButton-sizeSmall-149 TournamentCards-right-65 MuiCardActions-action-127"
        href="/tournament/mock1"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <span
          className="MuiButton-label-129"
        >
          Details
        </span>
        <span
          className="MuiTouchRipple-root-155"
        />
      </a>
    </div>
  </div>
  <div
    className="MuiPaper-root-67 MuiPaper-elevation1-70 MuiPaper-rounded-68 MuiCard-root-66 TournamentCards-card-64"
    id="mock2"
  >
    <div
      className="MuiCardHeader-root-94"
    >
      <div
        className="MuiCardHeader-content-97"
      >
        <span
          className="MuiTypography-root-100 MuiTypography-headline-105 MuiCardHeader-title-98"
        >
          test2
        </span>
        <span
          className="MuiTypography-root-100 MuiTypography-body1-109 MuiTypography-colorTextSecondary-123 MuiCardHeader-subheader-99"
        >
          cat0
        </span>
      </div>
    </div>
    <div
      className="MuiCardContent-root-125"
    >
      <p
        className="MuiTypography-root-100 MuiTypography-body1-109"
      >
        4 Total Players
      </p>
      <p
        className="MuiTypography-root-100 MuiTypography-body1-109"
      >
        4 Active Players
      </p>
    </div>
    <div
      className="MuiCardActions-root-126"
    >
      <a
        className="MuiButtonBase-root-152 MuiButton-root-128 MuiButton-text-130 MuiButton-flat-133 MuiButton-sizeSmall-149 TournamentCards-right-65 MuiCardActions-action-127"
        href="/new/mock2"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <span
          className="MuiButton-label-129"
        >
          Continue
        </span>
        <span
          className="MuiTouchRipple-root-155"
        />
      </a>
    </div>
  </div>
</div>
`;

exports[`matches the prior snapshot 3`] = `
<div
  className="TournamentCards-root-63"
>
  <div
    className="MuiPaper-root-67 MuiPaper-elevation1-70 MuiPaper-rounded-68 MuiCard-root-66 TournamentCards-card-64"
    id="mock0"
  >
    <div
      className="MuiCardHeader-root-94"
    >
      <div
        className="MuiCardHeader-content-97"
      >
        <span
          className="MuiTypography-root-100 MuiTypography-headline-105 MuiCardHeader-title-98"
        >
          test0
        </span>
        <span
          className="MuiTypography-root-100 MuiTypography-body1-109 MuiTypography-colorTextSecondary-123 MuiCardHeader-subheader-99"
        >
          cat0
        </span>
      </div>
    </div>
    <div
      className="MuiCardContent-root-125"
    >
      <p
        className="MuiTypography-root-100 MuiTypography-body1-109"
      >
        4 Total Players
      </p>
      <p
        className="MuiTypography-root-100 MuiTypography-body1-109"
      >
        3 Active Players
      </p>
    </div>
  </div>
  <div
    className="MuiPaper-root-67 MuiPaper-elevation1-70 MuiPaper-rounded-68 MuiCard-root-66 TournamentCards-card-64"
    id="mock1"
  >
    <div
      className="MuiCardHeader-root-94"
    >
      <div
        className="MuiCardHeader-content-97"
      >
        <span
          className="MuiTypography-root-100 MuiTypography-headline-105 MuiCardHeader-title-98"
        >
          test1
        </span>
        <span
          className="MuiTypography-root-100 MuiTypography-body1-109 MuiTypography-colorTextSecondary-123 MuiCardHeader-subheader-99"
        >
          cat1
        </span>
      </div>
    </div>
    <div
      className="MuiCardContent-root-125"
    >
      <p
        className="MuiTypography-root-100 MuiTypography-body1-109"
      >
        8 Total Players
      </p>
      <p
        className="MuiTypography-root-100 MuiTypography-body1-109"
      >
        6 Active Players
      </p>
    </div>
  </div>
  <div
    className="MuiPaper-root-67 MuiPaper-elevation1-70 MuiPaper-rounded-68 MuiCard-root-66 TournamentCards-card-64"
    id="mock2"
  >
    <div
      className="MuiCardHeader-root-94"
    >
      <div
        className="MuiCardHeader-content-97"
      >
        <span
          className="MuiTypography-root-100 MuiTypography-headline-105 MuiCardHeader-title-98"
        >
          test2
        </span>
        <span
          className="MuiTypography-root-100 MuiTypography-body1-109 MuiTypography-colorTextSecondary-123 MuiCardHeader-subheader-99"
        >
          cat0
        </span>
      </div>
    </div>
    <div
      className="MuiCardContent-root-125"
    >
      <p
        className="MuiTypography-root-100 MuiTypography-body1-109"
      >
        4 Total Players
      </p>
      <p
        className="MuiTypography-root-100 MuiTypography-body1-109"
      >
        4 Active Players
      </p>
    </div>
  </div>
</div>
`;
